src/Npgsql/NpgsqlConnector.cs

```diff
@@ -775,6 +837,99 @@ void SetSocketOptions(Socket socket)
 
         #endregion
 
+        #region I/O
+
+        internal readonly ChannelReader<NpgsqlCommand>? CommandsInFlightReader;
+        internal readonly ChannelWriter<NpgsqlCommand>? CommandsInFlightWriter;
+
+        internal volatile int CommandsInFlightCount;
+
+        internal ManualResetValueTaskSource<object?> ReaderCompleted { get; } =
+            new ManualResetValueTaskSource<object?> { RunContinuationsAsynchronously = true };
+
+        async Task ReadLoop()
+        {
+            Debug.Assert(Settings.Multiplexing);
+            Debug.Assert(CommandsInFlightReader != null);
+
+            NpgsqlCommand? command = null;
+
+            try
+            {
+                while (await CommandsInFlightReader.WaitToReadAsync())
+                {
+                    var commandsRead = 0;
+                    Debug.Assert(!InTransaction);
+
+                    while (CommandsInFlightReader.TryRead(out command))
+                    {
+                        commandsRead++;
+
+                        // TODO: Don't forget to remove this (at least for net5.0)
+                        await ReadBuffer.Ensure(5, true);
+
+                        // TODO: async message (notifications/notices)
+
+                        // We have a resultset for the command - hand back control to the command (which will
+                        // return it to the user)
+                        ReaderCompleted.Reset();
+                        command.ExecutionCompletion.SetResult(this);
+
+                        // Now wait until that command's reader is disposed. Note that RunContinuationsAsynchronously is
+                        // true, so that the user code calling NpgsqlDataReader.Dispose will not continue executing
+                        // synchronously here. The prevents issues if the code after the next command's execution
+                        // completion blocks.
+                        await new ValueTask(ReaderCompleted, ReaderCompleted.Version);
+                        Debug.Assert(!InTransaction);
+                    }
+
+                    // Atomically update the commands in-flight counter, and check if it reached 0. If so, the
+                    // connector is idle and can be returned.
+                    // Note that this is racing with over-capacity writing, which can select any connector at any
+                    // time (see MultiplexingWriteLoop), and we must make absolutely sure that if a connector is
+                    // returned to the pool, it is *never* written to unless properly dequeued from the Idle channel.
+                    if (Interlocked.Add(ref CommandsInFlightCount, -commandsRead) == 0)
+                        _pool!.Return(this);
+                }
+
+                Log.Trace("Exiting multiplexing read loop", Id);
```

Comment:
If ChannelReader.WaitToReadAsync returned false, that means the channel completed and there's nothing left inside - so CommandsInFlightCount should definitely be 0.

Returning returning to the pool... I don't think there's a bug here, but it could be a bit tricky. I'm not aware of a situation where the connector can be broken without there first being some commands in the channel; this means that WaitToReadAsync would go inside and the actual reading would throw. So I can't see how the while would ever exit without an exception.

But in any case will add an assertion.

Note that if a multiplexing connector is bound (e.g. transaction), and is then broken, then when the binding scope is ended the connector is properly returned to the pool, as always. So for bound connectors we don't rely on the ReadLoop to return the connector to the pool.

Reply:
I see, yeah assertion will help with expressing intent, even if we think today's implementation can't fail.